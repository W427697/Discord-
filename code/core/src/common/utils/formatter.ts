import semver from 'semver';
import { dedent } from 'ts-dedent';

type Prettier = typeof import('prettier');
type PrettierVersion = 3;

let prettierInstance: Prettier | undefined;
let prettierVersion: 3 | null = null;

const getPrettier = async (): Promise<
  { instance: undefined; version: null } | { instance: typeof import('prettier'); version: 3 }
> => {
  if (!prettierInstance) {
    try {
      prettierInstance = (await import('prettier')) as unknown as Prettier | undefined;
      prettierVersion = prettierInstance?.version
        ? (semver.major(prettierInstance.version) as PrettierVersion)
        : null;

      return {
        version: prettierVersion,
        instance: prettierInstance,
      } as any;
    } catch (err) {
      return {
        instance: undefined,
        version: null,
      };
    }
  }

  return {
    instance: prettierInstance,
    version: prettierVersion,
  } as any;
};

/**
 * Format the content of a file using prettier.
 * If prettier is not available in the user's project, it will fallback to use editorconfig settings if available and formats the file by a prettier-fallback.
 */
export async function formatFileContent(filePath: string, content: string): Promise<string> {
  try {
    const prettier = await getPrettier();

    switch (prettier.version) {
      case 3:
        const config = await prettier.instance.resolveConfig(filePath);

        if (!config || Object.keys(config).length === 0) {
          return await formatWithEditorConfig(filePath, content);
        }

        const result = await prettier.instance.format(content, {
          ...(config as any),
          filepath: filePath,
        });

        return result;
      case null:
      case undefined:
        return await formatWithEditorConfig(filePath, content);
      default:
        console.warn(dedent`
        Your prettier version ${
          (prettier as any).version
        } is not supported to format files which were edited by Storybook. 
        Please raise an issue on the Storybook GitHub repository. 
        Falling back to EditorConfig settings, if available.
        `);
        return await formatWithEditorConfig(filePath, content);
    }
  } catch (error) {
    return content;
  }
}

async function formatWithEditorConfig(filePath: string, content: string) {
  const prettier = await import('prettier');
  const config = await prettier.resolveConfig(filePath, { editorconfig: true });

  if (!config || Object.keys(config).length === 0) {
    return content;
  }

  return prettier.format(content, {
    ...(config as any),
    filepath: filePath,
  });
}
