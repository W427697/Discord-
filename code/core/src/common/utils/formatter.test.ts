import { formatFileContent } from './formatter';
import { describe, it, expect, vi } from 'vitest';
import path from 'node:path';

const mockPrettier = vi.hoisted(() => ({
  resolveConfig: vi.fn(),
  format: vi.fn(),
  version: vi.fn(),
}));

vi.mock('prettier', () => ({
  resolveConfig: mockPrettier.resolveConfig,
  format: mockPrettier.format,
  get version() {
    return mockPrettier.version();
  },
}));

const dummyContent = `
import type { Meta, StoryObj } from '@storybook/nextjs'

import Component from './foo';

  const meta = {
    component: Component
  } satisfies Meta<typeof Component>;

export default meta;

type Story = StoryObj<typeof meta>;
`;

describe('formatter', () => {
  describe('withPrettierConfig', () => {
    const testPath = path.resolve(__dirname, '__tests-formatter__', 'withPrettierConfig');

    describe('prettier', async () => {
      const prettierV3 = await import('prettier');

      it('formats content with prettier', async () => {
        mockPrettier.format.mockImplementation(prettierV3.format);
        mockPrettier.version.mockReturnValue(prettierV3.version);
        mockPrettier.resolveConfig.mockImplementation(prettierV3.resolveConfig);

        const filePath = path.resolve(testPath, 'testFile.ts');

        const result = await formatFileContent(filePath, dummyContent);

        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('withoutPrettierConfigAndWithEditorConfig', () => {
    const testPath = path.resolve(__dirname, '__tests-formatter__', 'withoutPrettierConfig');

    describe('prettier-v3', async () => {
      const prettierV3 = await import('prettier');

      it('formats content with prettier', async () => {
        mockPrettier.format.mockImplementation(prettierV3.format);
        mockPrettier.version.mockReturnValue(prettierV3.version);
        mockPrettier.resolveConfig.mockImplementation(prettierV3.resolveConfig);

        const filePath = path.resolve(testPath, 'testFile.ts');

        const result = await formatFileContent(filePath, dummyContent);

        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('withoutPrettierConfigAndWithEditorConfig', () => {
    const testPath = path.resolve(__dirname, '__tests-formatter__', 'withoutEditorConfig');

    describe('prettier-v3', async () => {
      const prettierV3 = await import('prettier');

      it('formats content with prettier', async () => {
        mockPrettier.format.mockImplementation(prettierV3.format);
        mockPrettier.version.mockReturnValue(prettierV3.version);
        mockPrettier.resolveConfig.mockResolvedValue(null);

        const filePath = path.resolve(testPath, 'testFile.ts');

        const result = await formatFileContent(filePath, dummyContent);

        expect(result).toBe(dummyContent);
      });
    });
  });
});
