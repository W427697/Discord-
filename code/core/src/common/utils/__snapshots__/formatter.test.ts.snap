// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatter > withPrettierConfig > prettier > formats content with prettier 1`] = `
"
import type { Meta, StoryObj } from '@storybook/nextjs'

import Component from './foo';

  const meta = {
    component: Component
  } satisfies Meta<typeof Component>;

export default meta;

type Story = StoryObj<typeof meta>;
"
`;

exports[`formatter > withoutPrettierConfigAndWithEditorConfig > prettier-v3 > formats content with prettier 1`] = `
"
import type { Meta, StoryObj } from '@storybook/nextjs'

import Component from './foo';

  const meta = {
    component: Component
  } satisfies Meta<typeof Component>;

export default meta;

type Story = StoryObj<typeof meta>;
"
`;
